File: crates/app/src/updater.rs
//! In-place app self-update.
//!
//! The app downloads the release archive itself (curl applies no quarantine
//! xattr), verifies the release asset's SHA-256 digest, stages the new bundle
//! next to the current one, and swaps them with two same-directory renames so
//! the operation is atomic per rename and can be rolled back. A detached
//! helper relaunches the new bundle after this process exits.

use std::{
    path::{Path, PathBuf},
    process::{Command, Stdio},
};

use serde_json::Value;

const DEFAULT_REPO: &str = "shivamhwp/wrec";
const CURRENT_VERSION: &str = env!("CARGO_PKG_VERSION");

#[derive(Clone, Debug, PartialEq, Eq)]
pub(crate) enum AppUpdateState {
    Idle,
    Checking,
    UpToDate,
    Available { version: String },
    Updating,
    Failed { message: String },
}

#[derive(Clone, Debug)]
pub(crate) struct ReadyUpdate {
    pub(crate) version: String,
    pub(crate) bundle: PathBuf,
    pub(crate) old_bundle: PathBuf,
}

/// The bundle this process runs from, when it is one the updater may replace:
/// a packaged, non-dev `.app`. Dev bundles and bare Cargo binaries update
/// through a rebuild instead.
pub(crate) fn eligible_bundle() -> Result<PathBuf, String> {
    let bundle = std::env::current_exe()
        .ok()
        .and_then(|exe| {
            exe.ancestors()
                .find(|path| path.extension().and_then(|ext| ext.to_str()) == Some("app"))
                .map(Path::to_path_buf)
        })
        .ok_or("not running from an app bundle; update by rebuilding instead")?;

    let is_dev = bundle
        .file_name()
        .and_then(|name| name.to_str())
        .is_some_and(|name| name.contains("Dev"));
    if is_dev {
        return Err("dev builds update by rebuilding, not in place".into());
    }
    Ok(bundle)
}

/// Returns `Some(version)` when a newer release exists, `None` when current.
pub(crate) fn check() -> Result<Option<String>, String> {
    eligible_bundle()?;
    let release = latest_release()?;
    Ok(is_newer(&release.version, CURRENT_VERSION).then_some(release.version))
}

pub(crate) fn download_and_apply() -> Result<ReadyUpdate, String> {
    let bundle = eligible_bundle()?;
    let release = latest_release()?;
    if !is_newer(&release.version, CURRENT_VERSION) {
        return Err(format!("wrec {CURRENT_VERSION} is already up to date"));
    }
    let asset = release
        .asset
        .as_ref()
        .ok_or_else(|| format!("release v{} has no app archive asset yet", release.version))?;

    let work_dir = exclusive_work_dir()?;
    let result = stage_and_swap(&bundle, asset, &work_dir, &release.version);
    let _ = std::fs::remove_dir_all(&work_dir);
    result
}

/// Relaunches the swapped-in bundle once this process exits and removes the
/// old bundle afterwards. Must be called right before quitting.
pub(crate) fn relaunch_and_cleanup(update: &ReadyUpdate) {
    let script = format!(
        "sleep 1; open -n \"{}\"; sleep 5; rm -rf \"{}\"",
        update.bundle.display(),
        update.old_bundle.display()
    );
    let _ = Command::new("sh")
        .args(["-c", &script])
        .stdin(Stdio::null())
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .spawn();
}

fn stage_and_swap(
    bundle: &Path,
    asset: &ReleaseAsset,
    work_dir: &Path,
    version: &str,
) -> Result<ReadyUpdate, String> {
    let archive = work_dir.join("wrec-app.tar.gz");
    download(&asset.url, &archive)?;

    let Some(expected) = &asset.sha256 else {
        return Err(
            "the release asset publishes no SHA-256 digest; refusing to install an unverifiable archive"
                .into(),
        );
    };
    let actual = sha256_of(&archive)?;
    if &actual != expected {
        return Err(format!(
            "checksum mismatch for the app archive: expected {expected}, got {actual}; not installing"
        ));
    }

    let extracted_dir = work_dir.join("extracted");
    std::fs::create_dir_all(&extracted_dir)
        .map_err(|err| format!("could not create {}: {err}", extracted_dir.display()))?;
    run_ok(
        Command::new("/usr/bin/tar")
            .args(["-xzf"])
            .arg(&archive)
            .arg("-C")
            .arg(&extracted_dir),
        "could not extract the app archive",
    )?;
    let new_bundle = find_app_bundle(&extracted_dir)?;

    // The daemon ships inside the bundle; stop it so the relaunched app
    // starts a matching one. Refusals (active recording) surface here.
    if let Err(err) = control::send_request("daemon.stop", serde_json::json!({})) {
        if err.code == "daemon_busy" {
            return Err(format!("{} {}", err.message, err.next));
        }
        // Unreachable daemon just means there is nothing to stop.
    }

    let old_bundle = swap_bundles(bundle, &new_bundle)?;
    Ok(ReadyUpdate {
        version: version.to_string(),
        bundle: bundle.to_path_buf(),
        old_bundle,
    })
}

/// Replaces `bundle` with `replacement` using a staged copy in the same
/// directory (same-volume renames are atomic). Returns where the old bundle
/// was moved; restores it if the final rename fails.
fn swap_bundles(bundle: &Path, replacement: &Path) -> Result<PathBuf, String> {
    let parent = bundle
        .parent()
        .ok_or("app bundle has no parent directory")?;
    let name = bundle
        .file_name()
        .and_then(|name| name.to_str())
        .ok_or("app bundle has no name")?;
    let pid = std::process::id();
    let staged = parent.join(format!(".{name}.staged-{pid}"));
    let old = parent.join(format!(".{name}.old-{pid}"));

    // ditto preserves permissions, symlinks, and signatures across volumes.
    run_ok(
        Command::new("/usr/bin/ditto").arg(replacement).arg(&staged),
        "could not stage the new app bundle (is the install directory writable?)",
    )?;

    if let Err(err) = std::fs::rename(bundle, &old) {
        let _ = std::fs::remove_dir_all(&staged);
        return Err(format!("could not move the current app aside: {err}"));
    }
    if let Err(err) = std::fs::rename(&staged, bundle) {
        let _ = std::fs::rename(&old, bundle);
        let _ = std::fs::remove_dir_all(&staged);
        return Err(format!("could not move the new app into place: {err}"));
    }
    Ok(old)
}

fn find_app_bundle(dir: &Path) -> Result<PathBuf, String> {
    let entries =
        std::fs::read_dir(dir).map_err(|err| format!("could not read {}: {err}", dir.display()))?;
    for entry in entries.flatten() {
        let path = entry.path();
        if path.extension().and_then(|ext| ext.to_str()) == Some("app") {
            let executable = path.join("Contents").join("MacOS").join("wrec-app");
            if executable.is_file() {
                return Ok(path);
            }
            return Err(format!(
                "archive bundle {} is missing Contents/MacOS/wrec-app",
                path.display()
            ));
        }
    }
    Err("the app archive contained no .app bundle".into())
}

#[derive(Debug)]
struct Release {
    version: String,
    asset: Option<ReleaseAsset>,
}

#[derive(Debug)]
struct ReleaseAsset {
    url: String,
    sha256: Option<String>,
}

fn latest_release() -> Result<Release, String> {
    let repo = std::env::var("WREC_REPO").unwrap_or_else(|_| DEFAULT_REPO.to_string());
    let url = format!("https://api.github.com/repos/{repo}/releases/latest");
    let output = Command::new("curl")
        .args([
            "-fsSL",
            "--retry",
            "3",
            "-H",
            "Accept: application/vnd.github+json",
            "-H",
            "User-Agent: wrec-app",
            &url,
        ])
        .output()
        .map_err(|err| format!("could not run curl: {err}"))?;
    if !output.status.success() {
        return Err(format!(
            "could not query the latest release: {}",
            String::from_utf8_lossy(&output.stderr).trim()
        ));
    }
    let body: Value = serde_json::from_slice(&output.stdout)
        .map_err(|err| format!("could not parse the release response: {err}"))?;
    parse_release(&body)
}

fn parse_release(body: &Value) -> Result<Release, String> {
    let tag = body
        .get("tag_name")
        .and_then(Value::as_str)
        .ok_or("release response had no tag_name")?;
    let version = tag.trim_start_matches('v').to_string();

    let asset_name = app_asset_name();
    let asset = body
        .get("assets")
        .and_then(Value::as_array)
        .and_then(|assets| {
            assets.iter().find(|entry| {
                entry.get("name").and_then(Value::as_str) == Some(asset_name.as_str())
            })
        })
        .and_then(|entry| {
            Some(ReleaseAsset {
                url: entry.get("browser_download_url")?.as_str()?.to_string(),
                sha256: entry
                    .get("digest")
                    .and_then(Value::as_str)
                    .and_then(|digest| digest.strip_prefix("sha256:"))
                    .map(str::to_string),
            })
        });

    Ok(Release { version, asset })
}

fn app_asset_name() -> String {
    let target = if cfg!(target_arch = "aarch64") {
        "aarch64-apple-darwin"
    } else {
        "x86_64-apple-darwin"
    };
    format!("wrec-app-{target}.tar.gz")
}

/// Treats an unparseable version pair as an update when the strings differ, so
/// a rename of the tagging scheme never strands installed apps.
fn is_newer(latest: &str, current: &str) -> bool {
    match (parse_version(latest), parse_version(current)) {
        (Some(latest), Some(current)) => latest > current,
        _ => latest != current,
    }
}

fn parse_version(version: &str) -> Option<(u64, u64, u64)> {
    let mut parts = version.trim_start_matches('v').splitn(3, '.');
    let major = parts.next()?.parse().ok()?;
    let minor = parts.next()?.parse().ok()?;
    // Ignore a pre-release suffix ("1-rc1" -> 1) so such tags compare by
    // their numeric patch instead of the strings-differ fallback.
    let patch = parts.next()?;
    let digits: String = patch.chars().take_while(char::is_ascii_digit).collect();
    Some((major, minor, digits.parse().ok()?))
}

/// A fresh, owner-only workspace whose creation fails rather than adopting an
/// existing directory, so nothing can pre-place files we later trust.
fn exclusive_work_dir() -> Result<PathBuf, String> {
    let nanos = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|elapsed| elapsed.as_nanos())
        .unwrap_or_default();
    let dir = std::env::temp_dir().join(format!("wrec-app-update-{}-{nanos}", std::process::id()));
    std::fs::create_dir(&dir)
        .map_err(|err| format!("could not create {}: {err}", dir.display()))?;
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let _ = std::fs::set_permissions(&dir, std::fs::Permissions::from_mode(0o700));
    }
    Ok(dir)
}

fn sha256_of(path: &Path) -> Result<String, String> {
    let output = Command::new("shasum")
        .args(["-a", "256"])
        .arg(path)
        .output()
        .map_err(|err| format!("could not run shasum: {err}"))?;
    if !output.status.success() {
        return Err(format!(
            "shasum failed: {}",
            String::from_utf8_lossy(&output.stderr).trim()
        ));
    }
    String::from_utf8_lossy(&output.stdout)
        .split_whitespace()
        .next()
        .map(str::to_string)
        .ok_or_else(|| "shasum produced no output".into())
}

fn download(url: &str, output: &Path) -> Result<(), String> {
    let result = Command::new("curl")
        .args(["-fsSL", "--retry", "3", "-o"])
        .arg(output)
        .arg(url)
        .stdout(Stdio::null())
        .stderr(Stdio::piped())
        .output()
        .map_err(|err| format!("could not run curl: {err}"))?;
    if !result.status.success() {
        return Err(format!(
            "download failed for {url}: {}",
            String::from_utf8_lossy(&result.stderr).trim()
        ));
    }
    Ok(())
}

fn run_ok(command: &mut Command, context: &str) -> Result<(), String> {
    let output = command
        .stdout(Stdio::null())
        .stderr(Stdio::piped())
        .output()
        .map_err(|err| format!("{context}: {err}"))?;
    if !output.status.success() {
        return Err(format!(
            "{context}: {}",
            String::from_utf8_lossy(&output.stderr).trim()
        ));
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn temp_dir(label: &str) -> PathBuf {
        let dir =
            std::env::temp_dir().join(format!("wrec-updater-test-{label}-{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).unwrap();
        dir
    }

    fn write_marker(bundle: &Path, contents: &str) {
        std::fs::create_dir_all(bundle).unwrap();
        std::fs::write(bundle.join("marker"), contents).unwrap();
    }

    #[test]
    fn version_ordering_detects_updates() {
        assert!(is_newer("0.2.1", "0.2.0"));
        assert!(is_newer("0.10.0", "0.9.9"));
        assert!(!is_newer("0.2.0", "0.2.0"));
        assert!(!is_newer("0.1.9", "0.2.0"));
        assert!(is_newer("2026.1", "0.2.0"));
    }

    #[test]
    fn swap_replaces_bundle_and_keeps_the_old_one() {
        let dir = temp_dir("swap");
        let bundle = dir.join("Wrec.app");
        let replacement = dir.join("New.app");
        write_marker(&bundle, "old");
        write_marker(&replacement, "new");

        let old = swap_bundles(&bundle, &replacement).unwrap();

        assert_eq!(
            std::fs::read_to_string(bundle.join("marker")).unwrap(),
            "new"
        );
        assert_eq!(std::fs::read_to_string(old.join("marker")).unwrap(), "old");
        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn swap_fails_cleanly_when_replacement_is_missing() {
        let dir = temp_dir("swap-missing");
        let bundle = dir.join("Wrec.app");
        write_marker(&bundle, "old");

        let err = swap_bundles(&bundle, &dir.join("Nope.app")).unwrap_err();

        assert!(err.contains("could not stage"));
        assert_eq!(
            std::fs::read_to_string(bundle.join("marker")).unwrap(),
            "old"
        );
        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn release_parsing_reads_tag_and_optional_asset() {
        let asset_name = app_asset_name();
        let body = serde_json::json!({
            "tag_name": "v0.3.0",
            "assets": [{
                "name": asset_name,
                "browser_download_url": "https://example.com/app.tar.gz",
                "digest": "sha256:abcd"
            }]
        });
        let release = parse_release(&body).unwrap();
        assert_eq!(release.version, "0.3.0");
        let asset = release.asset.unwrap();
        assert_eq!(asset.url, "https://example.com/app.tar.gz");
        assert_eq!(asset.sha256.as_deref(), Some("abcd"));

        let no_asset = serde_json::json!({ "tag_name": "v0.3.0", "assets": [] });
        assert!(parse_release(&no_asset).unwrap().asset.is_none());
    }

    #[test]
    fn app_bundle_discovery_requires_the_executable() {
        let dir = temp_dir("find");
        let bundle = dir.join("Wrec.app");
        std::fs::create_dir_all(bundle.join("Contents").join("MacOS")).unwrap();

        assert!(find_app_bundle(&dir).is_err());

        std::fs::write(
            bundle.join("Contents").join("MacOS").join("wrec-app"),
            "stub",
        )
        .unwrap();
        assert_eq!(find_app_bundle(&dir).unwrap(), bundle);
        let _ = std::fs::remove_dir_all(&dir);
    }
}

File: apps/mac/Sources/WrecApp/Core/Updater.swift
// Self-update: GitHub latest release → verified download → atomic bundle
// swap → relaunch. Port of `crates/app/src/updater.rs` with native APIs:
// URLSession instead of curl, CryptoKit streaming SHA-256 instead of shasum.
// Only `tar` and `ditto` remain as subprocesses (both are the right tool).

import CryptoKit
import Foundation

enum UpdateState: Equatable {
    case idle
    case checking
    case upToDate
    case available(String)
    case updating(String)
    case failed(String)
    case ineligible

    var buttonLabel: String {
        switch self {
        case .idle: "Check for updates"
        case .checking: "Checking…"
        case .upToDate: "Up to date"
        case .available(let version): "Update to \(version)"
        case .updating: "Updating…"
        case .failed: "Check again"
        case .ineligible: "Unavailable"
        }
    }

    var buttonDisabled: Bool {
        switch self {
        case .checking, .upToDate, .updating, .ineligible: true
        default: false
        }
    }
}

struct ReadyUpdate: Sendable {
    let version: String
    let bundle: URL
    let oldBundle: URL
}

enum UpdaterError: Error, CustomStringConvertible {
    case message(String)

    var description: String {
        switch self { case .message(let text): text }
    }
}

enum Updater {
    private static var repo: String {
        ProcessInfo.processInfo.environment["WREC_REPO"] ?? "shivamdoting/wrec"
    }

    static var currentVersion: String { Bundle.main.shortVersion }

    /// Update only a packaged, non-dev bundle — dev and `swift run` builds
    /// update by rebuilding.
    static func eligibleBundle() -> URL? {
        guard let bundle = Platform.currentAppBundle(), !Platform.isDevBundle() else {
            return nil
        }
        return bundle
    }

    struct Release: Sendable {
        let version: String
        let assetURL: URL
        let sha256: String?
    }

    static func check() async throws -> Release? {
        let release = try await latestRelease()
        guard isNewer(release.version, than: currentVersion) else { return nil }
        return release
    }

    static func downloadAndApply(_ release: Release, daemon: DaemonClient) async throws
        -> ReadyUpdate
    {
        guard let bundle = eligibleBundle() else {
            throw UpdaterError.message("this build updates by rebuilding")
        }

        let work = try exclusiveWorkDir()
        defer { try? FileManager.default.removeItem(at: work) }

        let archive = work.appending(path: "update.tar.gz")
        let (downloaded, _) = try await URLSession.shared.download(from: release.assetURL)
        try FileManager.default.moveItem(at: downloaded, to: archive)

        guard let expected = release.sha256 else {
            throw UpdaterError.message("release has no published digest; refusing to update")
        }
        let actual = try sha256(of: archive)
        guard actual == expected.lowercased() else {
            throw UpdaterError.message("digest mismatch: expected \(expected), got \(actual)")
        }

        let extracted = work.appending(path: "extracted")
        try FileManager.default.createDirectory(at: extracted, withIntermediateDirectories: true)
        try run("/usr/bin/tar", "-xzf", archive.path, "-C", extracted.path)
        let newBundle = try findAppBundle(in: extracted)
        try verifyReplacementBundle(newBundle, replacing: bundle)

        // Stop a running daemon so its binaries can be swapped. No daemon is
        // nothing to stop; only an active recording (daemon_busy) aborts.
        do {
            try await daemon.stopDaemon()
        } catch IPCError.unreachable {
            // Already stopped.
        } catch let IPCError.daemon(err) where err.code != "daemon_busy" {
            // Anything short of an active recording won't block the swap.
        }

        let oldBundle = try swapBundles(current: bundle, replacement: newBundle)
        return ReadyUpdate(version: release.version, bundle: bundle, oldBundle: oldBundle)
    }

    /// Detached: relaunch the swapped bundle, then reap the old one.
    static func relaunchAndCleanup(_ update: ReadyUpdate) {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/bin/sh")
        process.arguments = [
            "-c", #"sleep 1; open -n "$0"; sleep 5; rm -rf "$1""#,
            update.bundle.path, update.oldBundle.path,
        ]
        try? process.run()
    }

    // MARK: - GitHub

    private static func latestRelease() async throws -> Release {
        var request = URLRequest(
            url: URL(string: "https://api.github.com/repos/\(repo)/releases/latest")!)
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
        request.setValue("wrec-app", forHTTPHeaderField: "User-Agent")
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
            throw UpdaterError.message("release check failed")
        }

        struct Asset: Decodable {
            let name: String
            let browserDownloadUrl: URL
            let digest: String?
        }
        struct ReleaseJSON: Decodable {
            let tagName: String
            let assets: [Asset]
        }
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        let release = try decoder.decode(ReleaseJSON.self, from: data)

        let version = release.tagName.hasPrefix("v") ? String(release.tagName.dropFirst()) : release.tagName
        #if arch(arm64)
        let assetName = "wrec-app-aarch64-apple-darwin.tar.gz"
        #else
        let assetName = "wrec-app-x86_64-apple-darwin.tar.gz"
        #endif
        guard let asset = release.assets.first(where: { $0.name == assetName }) else {
            throw UpdaterError.message("release \(version) has no asset \(assetName)")
        }
        let sha = asset.digest.map { $0.hasPrefix("sha256:") ? String($0.dropFirst(7)) : $0 }
        return Release(version: version, assetURL: asset.browserDownloadUrl, sha256: sha)
    }

    static func isNewer(_ candidate: String, than current: String) -> Bool {
        if let a = parseVersion(candidate), let b = parseVersion(current) {
            for (x, y) in zip(a, b) where x != y { return x > y }
            return false
        }
        return candidate != current
    }

    /// `x.y.z` with trailing junk tolerated per component (`0.3.0-rc1` →
    /// `[0, 3, 0]`), so a suffixed tag equal to or older than the installed
    /// version is never offered as an update on every launch.
    private static func parseVersion(_ raw: String) -> [Int]? {
        let parts = raw.split(separator: ".", maxSplits: 2).map { Int($0.prefix(while: \.isNumber)) }
        guard parts.count == 3 else { return nil }
        let numbers = parts.compactMap { $0 }
        return numbers.count == 3 ? numbers : nil
    }

    // MARK: - Verify / swap

    private static func sha256(of file: URL) throws -> String {
        let handle = try FileHandle(forReadingFrom: file)
        defer { try? handle.close() }
        var hasher = SHA256()
        while let chunk = try handle.read(upToCount: 1 << 20), !chunk.isEmpty {
            hasher.update(data: chunk)
        }
        return hasher.finalize().map { String(format: "%02x", $0) }.joined()
    }

    private static func exclusiveWorkDir() throws -> URL {
        let dir = FileManager.default.temporaryDirectory
            .appending(path: "wrec-update-\(ProcessInfo.processInfo.processIdentifier)-\(UInt64.random(in: 0...UInt64.max))")
        try FileManager.default.createDirectory(
            at: dir, withIntermediateDirectories: false,
            attributes: [.posixPermissions: 0o700])
        return dir
    }

    private static func findAppBundle(in dir: URL) throws -> URL {
        let fm = FileManager.default
        let contents = try fm.contentsOfDirectory(at: dir, includingPropertiesForKeys: nil)
        for candidate in contents where candidate.pathExtension == "app" {
            let exe = candidate.appending(path: "Contents/MacOS/wrec-app")
            if fm.fileExists(atPath: exe.path) { return candidate }
        }
        throw UpdaterError.message("archive contains no wrec app bundle")
    }

    /// Reject malformed or differently identified bundles before the daemon
    /// is stopped. Ad-hoc releases fall back to the signing identifier; a
    /// Developer ID-signed app also pins replacements to its Apple team.
    static func verifyReplacementBundle(_ replacement: URL, replacing current: URL) throws {
        try run("/usr/bin/codesign", "--verify", "--deep", "--strict", replacement.path)

        let currentMetadata = try codesignMetadata(for: current)
        let replacementMetadata = try codesignMetadata(for: replacement)
        guard let currentIdentifier = currentMetadata["Identifier"],
            replacementMetadata["Identifier"] == currentIdentifier
        else {
            throw UpdaterError.message("update bundle has a different signing identifier")
        }

        if let currentTeam = teamIdentifier(in: currentMetadata) {
            guard currentTeam.count == 10,
                currentTeam.allSatisfy({ $0.isASCII && ($0.isUppercase || $0.isNumber) })
            else {
                throw UpdaterError.message("installed app has an invalid developer team identifier")
            }
            let requirement =
                "=anchor apple generic and certificate leaf[subject.OU] = \"\(currentTeam)\""
            do {
                try run(
                    "/usr/bin/codesign", "--verify", "--deep", "--strict",
                    "--test-requirement", requirement, replacement.path)
            } catch {
                throw UpdaterError.message("update bundle is not signed by the expected developer team")
            }
        }
    }

    private static func codesignMetadata(for bundle: URL) throws -> [String: String] {
        let process = Process()
        let output = Pipe()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/codesign")
        process.arguments = ["-d", "--verbose=4", bundle.path]
        process.standardOutput = output
        process.standardError = output
        try process.run()
        let data = output.fileHandleForReading.readDataToEndOfFile()
        process.waitUntilExit()
        guard process.terminationStatus == 0 else {
            throw UpdaterError.message("could not inspect update bundle signature")
        }

        return String(decoding: data, as: UTF8.self)
            .split(whereSeparator: \.isNewline)
            .reduce(into: [:]) { metadata, line in
                let parts = line.split(separator: "=", maxSplits: 1)
                if parts.count == 2 { metadata[String(parts[0])] = String(parts[1]) }
            }
    }

    private static func teamIdentifier(in metadata: [String: String]) -> String? {
        guard let value = metadata["TeamIdentifier"], value != "not set" else { return nil }
        return value
    }

    /// Stage with ditto (preserves signature), then two atomic same-dir
    /// renames. Rolls back if the final rename fails.
    private static func swapBundles(current: URL, replacement: URL) throws -> URL {
        let fm = FileManager.default
        let parent = current.deletingLastPathComponent()
        let name = current.lastPathComponent
        let pid = ProcessInfo.processInfo.processIdentifier
        let staged = parent.appending(path: ".\(name).staged-\(pid)")
        let old = parent.appending(path: ".\(name).old-\(pid)")

        try run("/usr/bin/ditto", replacement.path, staged.path)
        try fm.moveItem(at: current, to: old)
        do {
            try fm.moveItem(at: staged, to: current)
        } catch {
            try? fm.moveItem(at: old, to: current)
            try? fm.removeItem(at: staged)
            throw error
        }
        return old
    }

    private static func run(_ tool: String, _ args: String...) throws {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: tool)
        process.arguments = args
        try process.run()
        process.waitUntilExit()
        guard process.terminationStatus == 0 else {
            throw UpdaterError.message(
                "\(URL(fileURLWithPath: tool).lastPathComponent) exited \(process.terminationStatus)")
        }
    }
}
